### PR TITLE
PLF-8379 : remove disabled users from members list of the space

### PR DESCRIPTION
**Pull request by waelghannem**
remove desabled users from members list of the space.

### CODE REVIEW

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 1]
```diff
@@ -53,7 +53,13 @@
         @NamedQuery(name = "SpaceMember.getSpaceMembersByStatus",
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: Le nom de la requête est à modifier `SpaceMember.getSpaceMembersByStatus` to `SpaceMember.getEnabledSpaceMembersByStatus`

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 2]
```diff
@@ -53,7 +53,13 @@
...
                 query = "SELECT spaceMember.userId FROM SocSpaceMember AS spaceMember "
                         + " WHERE spaceMember.status = :status "
-                        + " AND   spaceMember.space.id = :spaceId "),
+                        + " AND   spaceMember.space.id = :spaceId "
+                        + " AND   (spaceMember.userId IN ( "
+                        + " SELECT identity.remoteId FROM SocIdentityEntity AS identity"
+                        + " WHERE  identity.enabled = TRUE "
+                        + " AND     identity.deleted = FALSE )"
+                        + " OR   spaceMember.userId NOT IN ( "
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: This condition should not be added to avoid displaying deleted users or users with no identity

File: component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/updater/SpaceMemberAnyMembershipUpgradeTest.java [hunk 3]
```diff
@@ -90,6 +92,7 @@ public void testSingleSpaceUpgrade() throws Exception {
     Space space = createSpace(spaceName);
     tearDownSpaceList.add(space);
     String[] members = space.getMembers();
+    LOG.info("members list size is {}", members.length);
```
Comment (boubaker) [inline component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/updater/SpaceMemberAnyMembershipUpgradeTest.java]: to delete

File: webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java [hunk 4]
```diff
@@ -256,7 +257,7 @@ public void setManagerList(List<Identity> managerList) {
     int elementsToDisplay = currentLoadManagerIndex + MANAGER_PER_PAGE;
     setManagerList(loadPeople(0, elementsToDisplay, Type.MANAGER));
     setEnableLoadManagerNext(elementsToDisplay < managerNum);
-    return managerList;
+    return managerList.stream().filter(identity -> identity.isEnable()).collect(Collectors.toList());
```
Comment (boubaker) [inline webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java]: to delete

File: webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java [hunk 5]
```diff
@@ -234,7 +235,7 @@ public void setMemberList(List<Identity> memberList) {
     int elementsToDisplay = currentLoadIndex + MEMBER_PER_PAGE;
     setMemberList(loadPeople(0, elementsToDisplay, Type.MEMBER));
     setEnableLoadNext(elementsToDisplay < memberNum);
-    return memberList;
+    return memberList.stream().filter(identity -> identity.isEnable()).collect(Collectors.toList());
```
Comment (boubaker) [inline webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java]: to delete

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 1 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/updater/SpaceMemberAnyMembershipUpgradeTest.java (hunk 3 above)
Comment (waelghannem) [reply component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/updater/SpaceMemberAnyMembershipUpgradeTest.java]: done.

File: webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java (hunk 5 above)
Comment (waelghannem) [reply webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java]: done.

File: webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java (hunk 4 above)
Comment (waelghannem) [reply webapp/portlet/src/main/java/org/exoplatform/social/portlet/UIMembersPortlet.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 2 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 6]
```diff
@@ -50,14 +50,22 @@
...
                         + "     spaceMember.status = :status "
                         + "   ) "),
-        @NamedQuery(name = "SpaceMember.getSpaceMembersByStatus",
+        @NamedQuery(name = "SpaceMember.getEnabledSpaceMembersByStatus",
                 query = "SELECT spaceMember.userId FROM SocSpaceMember AS spaceMember "
                         + " WHERE spaceMember.status = :status "
-                        + " AND   spaceMember.space.id = :spaceId "),
+                        + " AND   spaceMember.space.id = :spaceId "
+                        + " AND   spaceMember.userId IN ( "
+                        + " SELECT identity.remoteId FROM SocIdentityEntity AS identity"
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: We can't use this kind of inner queries because it's performance killer.
Can you measure this on database with a lot of users?

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 7]
```diff
@@ -50,14 +50,22 @@
...
+                        + " WHERE  identity.enabled = TRUE "
+                        + " AND     identity.deleted = FALSE )"),
         @NamedQuery(name = "SpaceMember.countSpaceMembersByStatus",
                 query = "SELECT count(*) FROM SocSpaceMember AS spaceMember "
                         + " WHERE spaceMember.status = :status "
-                        + " AND   spaceMember.space.id = :spaceId "),
+                        + " AND   spaceMember.space.id = :spaceId "
+                        + " AND   spaceMember.userId IN ( "
+                        + " SELECT identity.remoteId FROM SocIdentityEntity AS identity"
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: same as previous comment

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 6 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 7 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 8]
```diff
@@ -50,14 +50,20 @@
                         + "     spaceMember.userId = :userId AND "
                         + "     spaceMember.status = :status "
                         + "   ) "),
-        @NamedQuery(name = "SpaceMember.getSpaceMembersByStatus",
-                query = "SELECT spaceMember.userId FROM SocSpaceMember AS spaceMember "
+        @NamedQuery(name = "SpaceMember.getEnabledSpaceMembersByStatus",
+                query = "SELECT spaceMember.userId FROM SocSpaceMember AS spaceMember, SocIdentityEntity AS identity "
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: This query could be very slow, could you make some tests and return here found results to test its impact

File: component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml [hunk 9]
```diff
@@ -271,7 +271,19 @@
       </column>
       <column name="LAST_ACCESS" type="TIMESTAMP"/>
       <column name="VISITED" type="BOOLEAN"/>
-    </createTable>    
+    </createTable>
+      <createIndex tableName="SOC_SPACES_MEMBERS" indexName="IDX_SOC_SPACES_MEMBERS_01">
```
Comment (boubaker) [inline component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml]: The index creation should be made in a different changeSet to not impact existing Databases and to be applied correctly.

File: component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml [hunk 10]
```diff
@@ -308,6 +320,12 @@
         <constraints nullable="false"/>
       </column>
     </createTable>
+      <createIndex tableName="SOC_IDENTITIES" indexName="IDX_SOC_IDENTITIES_01">
```
Comment (boubaker) [inline component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml]: Same here

File: component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml [hunk 11]
```diff
@@ -271,7 +271,19 @@
       </column>
```
Comment (boubaker) [inline component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml]: Could you explain why the usage of these indexes specifically please. (FYI, we can combine multiple columns in index)

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 12]
```diff
@@ -50,14 +50,20 @@
...
+                        + " AND   spaceMember.userId = identity.remoteId "
+                        + " AND   identity.deleted = FALSE "
+                        + " AND   identity.enabled = TRUE "),
         @NamedQuery(name = "SpaceMember.countSpaceMembersByStatus",
-                query = "SELECT count(*) FROM SocSpaceMember AS spaceMember "
+                query = "SELECT count(*) FROM SocSpaceMember AS spaceMember, SocIdentityEntity AS identity "
                         + " WHERE spaceMember.status = :status "
-                        + " AND   spaceMember.space.id = :spaceId "),
+                        + " AND   spaceMember.space.id = :spaceId "
+                        + " AND   spaceMember.userId = identity.remoteId "
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: IMO, we should use "`INNER JOIN`" with "ON" condition using `spaceMember.userId = identity.remoteId`
In addition, to be able to use inner join, you will need to convert this query to SQL instead of JPQL.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java [hunk 13]
```diff
@@ -50,14 +50,20 @@
                         + "     spaceMember.userId = :userId AND "
                         + "     spaceMember.status = :status "
                         + "   ) "),
-        @NamedQuery(name = "SpaceMember.getSpaceMembersByStatus",
-                query = "SELECT spaceMember.userId FROM SocSpaceMember AS spaceMember "
+        @NamedQuery(name = "SpaceMember.getEnabledSpaceMembersByStatus",
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: No need to this query anymore, isn't it ?

File: component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java [hunk 14]
```diff
@@ -178,4 +182,82 @@ public void testGetLastAccessedPagination() throws Exception {
...
+     * There is 2 managers including 1 disabled manager in the store
+     * space managers list should not contain disabled managers **/
+    assertEquals(1, spaceFromCache.getManagers().length);
+  }
+
+  private Space createAndInitSpace(String remoteID, Boolean isDeleted, Boolean isDisabled) {
+    Space space = getSpaceInstance(1);
+    resetSpaceIdentitesStatus(space);
+    Identity spaceMemberIdentity = identityManager.getOrCreateIdentity(OrganizationIdentityProvider.NAME, remoteID, true);
+    if (isDisabled) spaceMemberIdentity.setEnable(!isDisabled);
```
Comment (boubaker) [inline component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java]: Use braces on conditions and loops please (DEVSTD-17)

File: component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java [hunk 15]
```diff
@@ -178,4 +182,82 @@ public void testGetLastAccessedPagination() throws Exception {
...
+    if (isDeleted) spaceMemberIdentity.setDeleted(isDeleted);
```
Comment (boubaker) [inline component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java]: Use braces on conditions and loops please (DEVSTD-17)

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java [hunk 16]
```diff
@@ -107,10 +107,16 @@ public void deleteBySpace(SpaceEntity entity) {
     if (limit <= 0) {
       throw new IllegalArgumentException("limit must be > 0");
     }
-    TypedQuery<String> query = getEntityManager().createNamedQuery("SpaceMember.getSpaceMembersByStatus",
-                                                                   String.class);
-    query.setParameter("status", status);
-    query.setParameter("spaceId", spaceId);
+
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: You didn't tested the result of `getSpaceMemberIdentitiesByProfileFilter` method after enbaling / disabling a space member. In fact, I think you didn't cleared the cache of `CachedIdentityStorage.identitiesCache` and `CachedIdentityStorage.identitiesCountCache` (with keys of type `ListSpaceMembersKey`) after enabling / disabling / deleting a user.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java [hunk 17]
```diff
@@ -124,10 +130,16 @@ public int countSpaceMembers(Long spaceId, SpaceMemberEntity.Status status) {
...
       throw new IllegalArgumentException("spaceId is null or equals to 0");
     }
-    TypedQuery<Long> query = getEntityManager().createNamedQuery("SpaceMember.countSpaceMembersByStatus", Long.class);
-    query.setParameter("status", status);
-    query.setParameter("spaceId", spaceId);
-    return query.getSingleResult().intValue();
+    StringBuilder queryStringBuilder = new StringBuilder("SELECT count(*) \n");
+    queryStringBuilder.append(" FROM SOC_SPACES_MEMBERS spaceMember \n");
+    queryStringBuilder.append(" INNER JOIN SOC_IDENTITIES identity \n");
+    queryStringBuilder.append("   ON spaceMember.USER_ID = identity.REMOTE_ID \n");
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: You can integrate in the `ON` condition both conditions `AND identity.ENABLED = TRUE AND identity.DELETED = FALSE` because it's used to select row from joined table `SOC_IDENTITIES`

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 8 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml (hunk 9 above)
Comment (waelghannem) [reply component/core-jpa/src/main/resources/db/changelog/social-rdbms.db.changelog-1.0.0.xml]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 12 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java (hunk 13 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/entity/SpaceMemberEntity.java]: done.

File: component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java (hunk 15 above)
Comment (waelghannem) [reply component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java (hunk 17 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: done.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java (hunk 16 above)
Comment (waelghannem) [reply component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: test added.

File: component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java (hunk 14 above)
Comment (waelghannem) [reply component/core-jpa/src/test/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageTest.java]: done.
